OKR.md

```diff
@@ -0,0 +1,52 @@
+# Quarterly Objectives and Key Results
+
+We try to frame our ongoing work using a process based on quarterly Objectives and Key Results (OKRs). Objectives reflect outcomes that are challenging, but realistic. Results are tangible and measurable.
+
+## 2018 Q4
+
+**go-ipfs handles large datasets (1TB++) without a sweat**
+- `P0` - It takes less than 48 hours to transfer 1TB dataset over Fast Ethernet (100Mbps)
+- `P0` @hannahhoward - List a sharded directory with 1M entries over LAN in under 1 minute, with less than a second to the first entry.
+- `PX` - There is a prototype implementation of GraphSync for go-ipfs
+- `P0` @magik6k - There is a better and more performant datastore module (e.g Badger or better)
+- `P1` - Rewrite pinning data structure to support large pinsets and flexible pin modes
```

Comment:
@kevina interested in contributing to this - not sure about owning per say yet =]

Reply:
I spent a very small amount of time working on something related to pinning and so I'd be interested in contributing to this objective. I think with enough support maybe I can own it, but I'd need to discuss that with someone who knows more about the system and the OKR to determine whether or not that is reasonable. And I don't want to step on your toes @kevina, if you are motivated to own this work.

Reply:
Link to the previous discussion: https://github.com/ipfs/go-ipfs/pull/5474#discussion_r220530372

To clarify my interest.

I think this task will be made a lot easier with some architectural changes to the blockstore.  I wrote up my first draft of such changes in https://github.com/ipfs/go-ipfs/issues/5528.  I also think that we should use a snapshot based approach to G.C. and completely avoid the need for read and write locks (but with perhaps mutexes that will be held for short durations).  In order for that to happen we will need some additional changes.  Once I write up this proposal I would like it to be given some serious thought as I think it can solve a lot of problems we are having.

If we agree to these changes I will gladly take lead to help push it though, although completing it in Q4 may be to aggressive a strategy.

I get the general sense that others would generally like to avoid any architectural and instead attempt to solve this with an independent data structures.  I personally see this as creating additional complexity so I not sure I can agree with this approach.  Thus, if we go down that path I probably not a good person to lead this.

Reply:
So, there are *two* issues here:

1. The KR is really about the pin datastructures itself. That is, we need to be able to (a) support flexible pin modes and (b) support *many* pins.
2. GC needs to get faster. A generational approach may work, personally, I'd go with a recounting approach. Regardless, I consider that to be an orthogonal issue.